go/kbfs/data/data_types.go

```diff
@@ -0,0 +1,449 @@
+// one indirect pointer with an indirect DirectType [although if it
+// holds for one, it should hold for all], and all of its indirect
+// pointers must have Ver 3, by c).
+type Ver int
+
+const (
+	// FirstValidVer is the first value that is considered a
+	// valid data version. Note that the nil value is not
+	// considered valid.
+	FirstValidVer Ver = 1
+	// ChildHolesVer is the data version for any indirect block
+	// containing a set of pointers with holes.
+	ChildHolesVer Ver = 2
+	// AtLeastTwoLevelsOfChildrenVer is the data version for
+	// blocks that have multiple levels of indirection below them
+	// (i.e., indirect blocks that point to other indirect blocks).
+	AtLeastTwoLevelsOfChildrenVer Ver = 3
+	// IndirectDirsVer is the data version for a directory block
+	// that contains indirect pointers.
+	IndirectDirsVer Ver = 4
+)
+
+// BlockReqType indicates whether an operation makes block
+// modifications or not
+type BlockReqType int
+
+const (
+	// BlockRead indicates a block read request.
+	BlockRead BlockReqType = iota
+	// BlockWrite indicates a block write request.
+	BlockWrite
+	// BlockReadParallel indicates a block read request that is
+	// happening from a different goroutine than the blockLock rlock
+	// holder, using the same lState.
+	BlockReadParallel
+	// BlockLookup indicates a lookup for a block for the purposes of
+	// creating a new node in the node cache for it; avoid any unlocks
+	// as part of the lookup process.
+	BlockLookup
+)
+
+// BlockDirectType indicates to what kind of block (direct or
+// indirect) a BlockPointer points.
+type BlockDirectType int
+
+const (
+	// UnknownDirectType indicates an old block that was written
+	// before we started labeling pointers.
+	UnknownDirectType BlockDirectType = 0
+	// DirectBlock indicates the pointed-to block has no indirect
+	// pointers.
+	DirectBlock BlockDirectType = 1
+	// IndirectBlock indicates the pointed-to block has indirect
+	// pointers.
+	IndirectBlock BlockDirectType = 2
+)
+
+func (bdt BlockDirectType) String() string {
+	switch bdt {
+	case UnknownDirectType:
+		return "unknown"
+	case DirectBlock:
+		return "direct"
+	case IndirectBlock:
+		return "indirect"
+	}
+	return fmt.Sprintf("<unknown blockDirectType %d>", bdt)
+}
+
+// BlockRef is a block ID/ref nonce pair, which defines a unique
+// reference to a block.
+type BlockRef struct {
+	ID       kbfsblock.ID
+	RefNonce kbfsblock.RefNonce
+}
+
+// IsValid returns true exactly when ID.IsValid() does.
+func (r BlockRef) IsValid() bool {
+	return r.ID.IsValid()
+}
+
+func (r BlockRef) String() string {
+	s := fmt.Sprintf("BlockRef{id: %s", r.ID)
+	if r.RefNonce != kbfsblock.ZeroRefNonce {
+		s += fmt.Sprintf(", refNonce: %s", r.RefNonce)
+	}
+	s += "}"
+	return s
+}
+
+// BlockPointer contains the identifying information for a block in KBFS.
+//
+// NOTE: Don't add or modify anything in this struct without
+// considering how old clients will handle them.
+type BlockPointer struct {
```

Comment:
I don't think there would be a circular dependency, but I think the main practical reasons are: 1) kbfsblock is used by the server, and `BlockPointer` is not needed by the server, so it would be a little wasteful, and 2) it's kind of weird for `kbfsblock` to depend on `kbfsmd`.

I was trying to keep `data` for strictly client-side stuff that the server doesn't use, but I could be convinced otherwise.

Reply:
I think the circular dependency would be in `Ver`, but yeah.

Reply:
Ah.  We could move `Ver` too if we moved it, and I think it would be ok.